### ISSUE TITLE
Shut down the server from within the serve method?

### ISSUE THREAD

**Issue by johann-petrak** [NanoHttpd/nanohttpd#355]
Is it possible to shut down the server as a result of a request from within the serve method? How?

**Comment by LordFokas** [comment]
call `this.stop()`, it will, *surprise surprise*, stop the HTTP/WS server :smile:

**Comment by johann-petrak** [comment]
I think there may be some problems with this approach: 
* the serve method should really send a response  to the request that triggers the shutdown, so if stop is called before the request is returned, that is not possible any more.
* the default is to keep the connection alive, so there will be an exception if the connection is closed using stop()
* but most importantly, at least in my case, the server will not stop after calling stop()  but hang. This is true for the server running either in demon or non-demon mode. This may be because I use the example code for using a thread pool, I still have to check this.

Currently I use the following work-around but it feels awfully like a hack:
* in the serve method, I start a thread that will call stop() and set a shutdown flag a little while after the response is returned
* the Connection header for the response is set to "close"
* the server is started in demon mode then the process is kept around by a loop that checks the shutdown flag periodically
* if the shutdown flag is set, I brutally System.exit(0), just letting the program run to the end would not terminate it

**Comment by LordFokas** [comment]
Alright then, let's make this about how to carefully stop accepting requests, then finish off any requests and proceed to kill the server socket.

It seems like a major behavior improvement so I'd also like to pull @ritchieGitHub into this one.

**Comment by ritchieGitHub** [comment]
Yes it would be nice if nanohttpd could be told to stop accepting connections and to automatically shutdown (and stop the thread pools) as soon as no connections are left. Probably we will not even need a thread to do this, but thats an implementation detail. This api should also be available from the outside, so to say a gracefullyStop()

**Comment by LordFokas** [comment]
Stopping gracefully should be the default behavior, but changing the behavior so much in that method (from killing the server to gracefully stopping) is a breaking change for apps that expect the server to be dead within a few cpu cycles.

I propose this:
- we deprecate `stop()` and add proper Javadocs with the new methods and behaviors. It should have a single line of code, invoking `kill()`.
- `kill()` now contains the old stop method, but properly documented and warning of the aggressive nature of the method.
- finally, insert the new behavior into the `terminate()` method that will now be the default method to stop the server. `NanoWSD` implementations should be forced to override this, or rather implement a callback invoked by this, to deal with any currently open websockets, while the default `NanoHTTPD` terminating behavior will still keep new websockets from being opened. This may lead to another question which is if or how to track currently open sockets but I think we can get over that one easily once we get to actually implement this.

**Comment by martinbonnin** [comment]
I'm using NanoHTTPD to create a small server to exchange a OAuth Token and that would be quite handy. As of now, there's a race and if users are unlucky, they get an error instead of a nice confirmation page.
